cluster-autoscaler/core/scaleup/orchestrator/executor.go

```diff
@@ -0,0 +1,240 @@
+		return err, extractNodeGroups(scaleUpInfos)
+	}
+	type errResult struct {
+		err  errors.AutoscalerError
+		info *nodegroupset.ScaleUpInfo
+	}
+	scaleUpsLen := len(scaleUpInfos)
+	errResults := make(chan errResult, scaleUpsLen)
+	var wg sync.WaitGroup
+	wg.Add(scaleUpsLen)
+	availableGPUTypes := e.autoscalingContext.CloudProvider.GetAvailableGPUTypes()
+	for _, scaleUpInfo := range scaleUpInfos {
+		go func(info nodegroupset.ScaleUpInfo) {
+			defer wg.Done()
+			nodeInfo, ok := nodeInfos[info.Group.Id()]
+			if !ok {
+				klog.Errorf("ExecuteScaleUp: failed to get node info for node group %s", info.Group.Id())
+				return
+			}
+			if aErr := e.executeScaleUp(info, nodeInfo, availableGPUTypes, now); aErr != nil {
+				errResults <- errResult{err: aErr, info: &info}
+			}
+		}(scaleUpInfo)
+	}
+	wg.Wait()
+	close(errResults)
+	var results []errResult
+	for err := range errResults {
+		results = append(results, err)
+	}
+	if len(results) > 0 {
+		failedNodeGroups := make([]cloudprovider.NodeGroup, len(results))
+		scaleUpErrors := make([]errors.AutoscalerError, len(results))
+		for i, result := range results {
+			failedNodeGroups[i] = result.info.Group
+			scaleUpErrors[i] = result.err
+		}
+		return combineConcurrentScaleUpErrors(scaleUpErrors), failedNodeGroups
+	}
+	return nil, nil
+}
+
+func (e *scaleUpExecutor) executeScaleUp(
+	info nodegroupset.ScaleUpInfo,
+	nodeInfo *schedulerframework.NodeInfo,
+	availableGPUTypes map[string]struct{},
+	now time.Time,
+) errors.AutoscalerError {
+	gpuConfig := e.autoscalingContext.CloudProvider.GetNodeGpuConfig(nodeInfo.Node())
+	gpuResourceName, gpuType := gpu.GetGpuInfoForMetrics(gpuConfig, availableGPUTypes, nodeInfo.Node(), nil)
+	klog.V(0).Infof("Scale-up: setting group %s size to %d", info.Group.Id(), info.NewSize)
+	e.autoscalingContext.LogRecorder.Eventf(apiv1.EventTypeNormal, "ScaledUpGroup",
+		"Scale-up: setting group %s size to %d instead of %d (max: %d)", info.Group.Id(), info.NewSize, info.CurrentSize, info.MaxSize)
+	increase := info.NewSize - info.CurrentSize
+	if err := info.Group.IncreaseSize(increase); err != nil {
+		e.autoscalingContext.LogRecorder.Eventf(apiv1.EventTypeWarning, "FailedToScaleUpGroup", "Scale-up failed for group %s: %v", info.Group.Id(), err)
+		aerr := errors.ToAutoscalerError(errors.CloudProviderError, err).AddPrefix("failed to increase node group size: ")
+		e.clusterStateRegistry.RegisterFailedScaleUp(info.Group, metrics.FailedScaleUpReason(string(aerr.Type())), now)
+		return aerr
+	}
+	e.clusterStateRegistry.RegisterOrUpdateScaleUp(
+		info.Group,
+		increase,
+		time.Now())
+	metrics.RegisterScaleUp(increase, gpuResourceName, gpuType)
+	e.autoscalingContext.LogRecorder.Eventf(apiv1.EventTypeNormal, "ScaledUpGroup",
+		"Scale-up: group %s size set to %d instead of %d (max: %d)", info.Group.Id(), info.NewSize, info.CurrentSize, info.MaxSize)
+	return nil
+}
+
+func combineConcurrentScaleUpErrors(errs []errors.AutoscalerError) errors.AutoscalerError {
```

Comment:
I thought about it as well. I didn't go that path because:
- this function is dedicated for combining concurrent errors (errors are deduplicated and sorted)
- it's based on an assumption that in this particular place we're ok with loosing some information about aggregated errors (as it's reported in [logs and state](https://github.com/kubernetes/autoscaler/pull/5731#discussion_r1190764705)). Thanks to this we omit changes in API.

...so there are some assumptions we're aware of. Having them in `errors` package could make this function the default way. WDYT?

Reply:
I think this is something that could be addressed by documenting in the comment and perhaps reflecting better in the function name (`Aggregate`? `Coalesce`? Naming is hard...) Leaving up to you though.